### PR TITLE
View for teaching assistants to navigate through source code 

### LINKED ISSUE #163: Create a View where teaching assistants can navigate through the source code
Currently, a teaching assistant can view the source code for an assignment only by downloading and opening the source files using an editor.

We want to have a GitHub like structure for each submission where the teaching assistant can view the source files from that submission.

Depends on #243

### PR DESCRIPTION
**Pull request by BiancaNecula**
Fixes #163 

The ta can see the source code or other files in submission directly in browser by accessing /submission/<id_of_submission>/<filename>.

### CODE REVIEW

File: interface/views.py [hunk 1]
```diff
@@ -327,3 +328,21 @@ def reveal(request, course_pk, assignment_pk):
...
     ))
+
+
+@login_required
+@staff_member_required
+def code_view(request, pk, filename):
+    submission = get_object_or_404(Submission, pk=pk)
+
+    if (request.user
+            not in submission.assignment.course.teaching_assistants.all()):
```
Comment (gmuraru) [inline interface/views.py:339]: Q: Might be worth creating a variable for this:
Ex:
```all_tas = submission.assignment.course.teaching_assistants.all()```
and then check:
```
if request_user not in all_tas:
```
Such that we do not split the if statement in two lines

File: interface/backend/minio_api.py [hunk 2]
```diff
@@ -31,6 +31,15 @@ def copy(source_file, destination_file):
                         f'/{settings.MINIO_BUCKET}/{source_file}')
 
 
+def download_buffer(filename, buffer):
+    try:
+        data = _client.get_object(settings.MINIO_BUCKET, filename)
+    except NoSuchKey:
+        raise MissingFile(filename)
+    for chunck in data.stream(32*1024):
```
Comment (gmuraru) [inline interface/backend/minio_api.py:39]: Q: Can we create a global variable for this value? (with uppercase letters)

File: interface/codeview.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import io
+import logging
+from zipfile import ZipFile
+
+from interface.backend.minio_api import MissingFile, download_buffer
+
+log = logging.getLogger(__name__)
+
+
+def submission_cut(request, submission, filename):
```
Comment (gmuraru) [inline interface/codeview.py:10]: Q: From what comes the name ```submission_cut```?

File: interface/codeview.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    try:
+        download_buffer(f'{submission.pk}.zip', buff)
+    except MissingFile:
+        return io.StringIO("The archive is missing!")
+
+    submission_archive = ZipFile(buff)
+    try:
+        file = submission_archive.read(
+            filename,
+            pwd=None,
```
Comment (Jokeswar) [inline interface/codeview.py:21]: `pwd` is, by default, set to `None`. No need to set it explicitly.

File: interface/codeview.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        )
+    except MissingFile:
```
Comment (Jokeswar) [inline interface/codeview.py:23]: I don't think `zipfile.read()` throws this kind of error. Let's leave it as generic as possible and remove `MissingFile` for the moment.

File: interface/views.py [hunk 6]
```diff
@@ -327,3 +328,21 @@ def reveal(request, course_pk, assignment_pk):
...
+        return HttpResponse(status=403)
+
+    file = submission_cut(request, submission, filename)
+
+    file_content = file.read()
+
+    file.close()
+    context = {'file_content': file_content}
```
Comment (Jokeswar) [inline interface/views.py:347]: Let's simplify this to something less spaced out. What do you say about:
```
context = {'file_content': file.read()}
file.close()
```

File: interface/codeview.py (hunk 3 above)
Comment (BiancaNecula) [reply interface/codeview.py:10]: It was the preliminary form of the function. It's actually `extract_file`.

File: interface/backend/minio_api.py [hunk 7]
```diff
@@ -7,6 +7,8 @@
 
 import io
 
+BLOCK_SIZE = 32*1024
```
Comment (Jokeswar) [inline interface/backend/minio_api.py:10]: Let's have this in settings.py

File: testsuite/test_student_workflow.py [hunk 8]
```diff
@@ -344,3 +344,30 @@ def test_user_see_revealed_score(client, STC, base_db_setup):
...
+            format='multipart',
+        )
+
+    submission = Submission.objects.all()[0]
+
+    response = client.get(
+        f'/submission/{submission.pk}/test',
+        follow=True,
+    )
+    STC.assertTemplateNotUsed(response, 'interface/code_view.html')
```
Comment (Jokeswar) [inline testsuite/test_student_workflow.py:373]: You could add a check for the response status and if it redirects to the admin page.

File: testsuite/test_student_workflow.py [hunk 9]
```diff
@@ -344,3 +344,30 @@ def test_user_see_revealed_score(client, STC, base_db_setup):
...
+    FILEPATH = settings.BASE_DIR / 'testsuite' / 'test.zip'
+
+    (_, _, user, course, assignment) = base_db_setup
+
+    client.login(username=user.username, password='pw')
+
+    with open(FILEPATH, 'rb') as file:
+        upload = SimpleUploadedFile(FILEPATH.name,
+                                    file.read(),
+                                    content_type='application/zip')
```
Comment (Jokeswar) [inline testsuite/test_student_workflow.py:360]: Let's go with a uniform coding style:
```
                                   # code ....
                                   content_type='application/zip',
)
```

File: testsuite/test_ta_workflow.py [hunk 10]
```diff
@@ -795,3 +798,31 @@ def test_ta_see_revealed_score(client, STC, base_db_setup):
...
+    FILEPATH = settings.BASE_DIR / 'testsuite' / 'test.zip'
+
+    (_, ta, _, course, assignment) = base_db_setup
+
+    client.login(username=ta.username, password='pw')
+
+    with open(FILEPATH, 'rb') as file:
+        upload = SimpleUploadedFile(FILEPATH.name,
+                                    file.read(),
+                                    content_type='application/zip')
```
Comment (Jokeswar) [inline testsuite/test_ta_workflow.py:814]: Same here as above

File: interface/codeview.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    buff = io.BytesIO()
+    try:
+        download_buffer(f'{submission.pk}.zip', buff)
+    except MissingFile:
+        return io.StringIO("The archive is missing!")
+
+    submission_archive = ZipFile(buff)
+    try:
+        file = submission_archive.read(filename)
+    except Exception:
```
Comment (Jokeswar) [inline interface/codeview.py:20]: The actual exception that we are looking for is `KeyError`. `Exception` is the father of all exceptions and would catch more than necessary.

File: interface/views.py [hunk 12]
```diff
@@ -327,3 +328,20 @@ def reveal(request, course_pk, assignment_pk):
...
     ))
+
+
+@login_required
+@staff_member_required
+def code_view(request, pk, filename):
+    submission = get_object_or_404(Submission, pk=pk)
+    all_tas = submission.assignment.course.teaching_assistants.all()
+
+    if (request.user not in all_tas):
```
Comment (gmuraru) [inline interface/views.py:339]: MAYBE: Remove parentheses

File: testsuite/test_student_workflow.py [hunk 13]
```diff
@@ -344,3 +344,38 @@ def test_user_see_revealed_score(client, STC, base_db_setup):
     )
     STC.assertNotContains(response, "N/A")
     STC.assertContains(response, "Final score")
+
+
+@pytest.mark.django_db
+def test_user_code_view(client, STC, base_db_setup):
+    FILEPATH = settings.BASE_DIR / 'testsuite' / 'test.zip'
```
Comment (gmuraru) [inline testsuite/test_student_workflow.py:351]: Q: Could you remove this? it is already defined at the top of this file

File: testsuite/test_ta_workflow.py [hunk 14]
```diff
@@ -795,3 +798,33 @@ def test_ta_see_revealed_score(client, STC, base_db_setup):
 
     response = client.get(f'/assignment/{course.pk}/{assignment.pk}')
     STC.assertNotContains(response, "N/A")
+
+
+@pytest.mark.django_db
+def test_ta_code_view(client, STC, base_db_setup):
+    FILEPATH = settings.BASE_DIR / 'testsuite' / 'test.zip'
```
Comment (gmuraru) [inline testsuite/test_ta_workflow.py:805]: Also here

**Pull request merged**